apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient3-plugin/src/main/java/co/elastic/apm/agent/httpclient/v3/HttpClient3Instrumentation.java

```diff
@@ -0,0 +1,138 @@
+import co.elastic.apm.agent.httpclient.HttpClientHelper;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Outcome;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContext;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.commons.httpclient.CircularRedirectException;
+import org.apache.commons.httpclient.HttpMethod;
+import org.apache.commons.httpclient.StatusLine;
+import org.apache.commons.httpclient.URI;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+/**
+ * Instruments {@link org.apache.commons.httpclient.HttpMethodDirector#executeMethod(HttpMethod)}
+ */
+@SuppressWarnings("JavadocReference") // instrumented class is package-private
+public class HttpClient3Instrumentation extends TracerAwareInstrumentation {
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("http-client", "apache-httpclient");
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return named("org.apache.commons.httpclient.HttpMethodDirector");
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("executeMethod");
+    }
+
+    @Override
+    public String getAdviceClassName() {
+        return "co.elastic.apm.agent.httpclient.v3.HttpClient3Instrumentation$HttpClient3Advice";
+    }
+
+    public static class HttpClient3Advice {
+
+        @Nullable
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object onEnter(@Advice.Argument(0) HttpMethod httpMethod) {
+            final AbstractSpan<?> parent = TracerAwareInstrumentation.tracer.getActive();
+            if (parent == null) {
+                return null;
+            }
+
+            String host;
+            String uri;
+            String protocol;
+            int port;
+            try {
+                URI httpClientURI = httpMethod.getURI();
+                host = httpClientURI.getHost();
+                uri = httpClientURI.toString();
+                protocol = httpClientURI.getScheme();
+                port = httpClientURI.getPort();
+            } catch (Exception e) {
+                // silently ignored
+                return null;
+            }
+
+            Span span = HttpClientHelper.startHttpClientSpan(parent, httpMethod.getName(), uri, protocol, host, port);
+
+            if (span != null) {
+                span.activate();
+            }
+
+            if (!TraceContext.containsTraceContextTextHeaders(httpMethod, HttpClient3RequestHeaderAccessor.INSTANCE)) {
```

Comment:
Outside of the scope of this PR but this seems to be a commonly repeated boilerplate. I wonder if it can be abstracted and re-used.

Reply:
I agree on this one, it's definitely duplicated more than once with a few variants. +1 on dealing with that in a separate PR.